lib/evmone/tracing.cpp

```diff
@@ -0,0 +1,51 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2021 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tracing.hpp"
+
+namespace evmone
+{
+namespace
+{
+class HistogramTracer : public Tracer
+{
+    evmc_revision m_rev{};
+    int32_t m_depth = 0;
+    const uint8_t* m_code = nullptr;
+    uint32_t m_counts[256]{};
+    std::ostream& m_out;
+
+    void on_execution_start(
+        evmc_revision rev, const evmc_message& msg, bytes_view code) noexcept override
+    {
+        m_rev = rev;
+        m_depth = msg.depth;
+        m_code = code.data();
+    }
+
+    void on_instruction_start(uint32_t pc) noexcept override { ++m_counts[m_code[pc]]; }
+
+    void on_execution_end(const evmc_result& /*result*/) noexcept override
+    {
+        const auto names = evmc_get_instruction_names_table(m_rev);
+        m_out << "--- # HISTOGRAM depth=" << m_depth << "\nopcode,count\n";
+        for (size_t i = 0; i < std::size(m_counts); ++i)
+        {
+            if (m_counts[i] != 0)
```

Comment:
Should we have thre options?
- do not print zeroes
- one which also prints all the zero opcodes
- only those zero opcodes which are defined for the evm version

Reply:
Some of these options make sense, but I don't think it is worth doing these right now.